### ISSUE TITLE
IRDL support for custom dialect definition

### ISSUE THREAD

**Issue by Moxinilian** [mlir-rs/melior#540]
I am currently adding [IRDL support for custom dialect definition to the C API](https://github.com/llvm/llvm-project/pull/91852), and once this has landed I'd like to add it to melior as well. Ideally a macro to generate useful Rust types from IRDL definitions would be really nice, but simply allowing to register custom dialects specified in IRDL would already be sufficient.

**Comment by Moxinilian** [comment]
As an update: I have managed to make this work on my fork of melior that uses LLVM 19. I assume there is no interest in LLVM 19-only features for now, so once LLVM 19 is released and supported in melior, I'll be able to submit a PR.

**Comment by TyrsDev** [comment]
This sounds excellent, @Moxinilian. I'm hoping for custom dialect support in melior soon. Do you have any resources or documentation available where I could see how this would be used in melior?

**Comment by Moxinilian** [comment]
Yes, basically it's just a matter of exposing [mlirLoadIRDLDialects](https://mlir.llvm.org/doxygen/mlir-c_2Dialect_2IRDL_8h.html#a6e95db2a616a1e4b8589911114d302ee), which takes a ModuleOp and loads any IRDL dialect defined in it.

**Comment by ProfFan** [comment]
LLVM 19.1.0 is out, https://github.com/llvm/llvm-project/releases/tag/llvmorg-19.1.0

**Comment by edg-l** [comment]
Made this in #640

**Comment by edg-l** [comment]
https://mlir-rs.github.io/melior/melior/dialect/ods/irdl/index.html
